Add java

MatMoore opened the pull request:
This is my first draft of the Java port, which uses maven to manage dependencies.

Differences from the python version:

- I didn't create an entry point for the program (a class with a main() method) because it seems like you can do the exercise with just classes and tests
- Added a getBalance() method because java doesn't have properties

Before merging this I'm going to see if I can bundle gradle as well (to give the candidate a choice) and check that other people besides me can build and run this.

MatMoore committed:
Add a java version of the test

This is the first test, using Junit 5 and targeting Java 8 for the
language level.

Tested with IntelliJ and Maven

MatMoore committed:
Get rid of some maven warnings

- Specify the encoding so build is not platform dependent
(https://maven.apache.org/general.html#encoding-warning)
- Specify a version for the compiler plugin

MatMoore committed:
Add remaining tests for Java version

MatMoore committed:
Ignore intellij compiled classes

MatMoore committed:
Add getting started steps (assuming maven is used)

exonian commented on java/src/test/java/OysterCardTest.java:
```diff
@@ -0,0 +1,43 @@
+
+    card.touchIn("TUBE", "Earl's Court");
+
+    assertEquals(card.getBalance(), 680);
```
Might these assertions be better the other way around? The failure messages I'm seeing are like this:
```org.opentest4j.AssertionFailedError: expected: <-1> but was: <680>```

3 options from here:
1) it's valuable to have these slightly confusing, cos that's life; 2) it's valuable to have them as clear as possible; 3) it really doesn't matter

MatMoore replied:
Good point! I didn't notice this. I think it's valuable to have them as clear as possible - interviews are stressful enough already so I don't think we should make the test harder that it needs to be

exonian replied:
100% agree, but then overthought it 😁

exonian reviewed (commented):
Nice! I brew installed maven and it was utterly plain sailing from there! Lovely stuff

One comment, about a thing which probably confused me more than it might someone actually doing the test...

MatMoore committed:
Add gradle

This provides another option for running the tests in case candidates are
more used to gradle (and would like to change the build configuration)

MatMoore committed:
Alter assertions to use the correct order for expected/actual values

exonian commented:
Thanks @MatMoore! 🎉
